lms/djangoapps/certificates/tests/test_tasks.py

```diff
@@ -32,4 +40,21 @@ def test_cert_task_missing_args(self, missing_arg):
 
         with patch('lms.djangoapps.certificates.tasks.User.objects.get'):
             with self.assertRaisesRegexp(KeyError, missing_arg):
-                generate_certificate(**kwargs)
+                generate_certificate.apply_async(kwargs=kwargs).get()
+
+    @patch('lms.djangoapps.certificates.tasks.generate_certificate.retry')
+    @patch('lms.djangoapps.certificates.tasks.generate_user_certificates')
+    def test_cert_task_retry(self, generate_user_certs_mock, mock_retry):
+        course_key = 'course-v1:edX+CS101+2017_T2'
+        student = UserFactory()
+
+        kwargs = {
+            'student': student.id,
+            'course_key': course_key,
+            'expected_verification_status': 'approved'
+        }
+        generate_certificate.apply_async(kwargs=kwargs)
+
+        SoftwareSecurePhotoVerificationFactory.create(user=student, status='approved')
+
+        self.assertTrue(mock_retry.called)
```

Comment:
Rather than mocking the call to `retry`, can you verify that the task was called twice?  This would catch if there are any issues to calling/raising `retry`.

Reply:
I don't think there is any way to check that a method was called without mocking it, but we don't want to mock the generate_certificate method because we want it to actually execute the first time. Is there another way we could check that the method was called? It doesn't return anything, so we can't check for return values.

FYI @iloveagent57

Reply:
Didn't you have an assertion in here previously about the call to `generate_user_certificates()`?  I think what you want to do is leave that mocked, don't mock out retry, but instead mock `SoftwareSecurePhotoVerification.user_status` to return a different status from what's passed in on the first call, and then the same status as what's passed in on the second call.  You can use `side_effect` to do this:
https://docs.python.org/3/library/unittest.mock.html (search for "side_effect").

Then assert that `generate_user_certificates()` is called exactly once with the expected arguments.